### PR TITLE
Tabular: Feature Importance

### PR DESCRIPTION
**Pull request by Innixma**
*Issue #, if available:*

#304 

*Description of changes:*

Major Update
- Added get_feature_importance() function to learner and trainer.
- Note: This is not surfaced to the user-facing API yet

Code Example:

```
import datetime
import pandas as pd

import sklearn.datasets

from autogluon.utils.tabular.ml.utils import generate_train_test_split
from autogluon import TabularPrediction as task

LABEL = 'target'


if __name__ == '__main__':
    utcnow = datetime.datetime.utcnow()
    timestamp_str_now = utcnow.strftime("%Y%m%d_%H%M%S")
    path_prefix = 'datasets/breast_cancer/'
    path_model_prefix = path_prefix + 'learners/' + timestamp_str_now + '/'

    cancer_data = sklearn.datasets.load_breast_cancer(return_X_y=False)
    X = pd.DataFrame(cancer_data['data'], columns=cancer_data['feature_names'])
    X[LABEL] = cancer_data['target']

    X_train, X_test, y_train, y_test = generate_train_test_split(X=X.drop(LABEL, axis=1), y=X[LABEL], problem_type='binary', test_size=0.2)

    X = X_train
    X[LABEL] = y_train
    X_test[LABEL] = y_test

    X = X.reset_index(drop=True)
    X_test = X_test.reset_index(drop=True)

    predictor = task.fit(
        train_data=X, label=LABEL, output_directory=path_model_prefix,
        hyperparameters={
            'RF': {'n_estimators': 10},
            'GBM': {'num_boost_round': 10},
            'CAT': {'iterations': 10},
        },
        auto_stack=True,
        eval_metric='roc_auc',
        enable_fit_continuation=True,
        verbosity=2,
    )

    print('CatboostClassifier_STACKER_l0')
    feature_importance = predictor._learner.get_feature_importance(model='CatboostClassifier_STACKER_l0')
    print(feature_importance)
    print()
    print('weighted_ensemble_k0_l1')
    feature_importance = predictor._learner.get_feature_importance(model='weighted_ensemble_k0_l1')
    print(feature_importance)


```
Code Output:

```
CatboostClassifier_STACKER_l0
worst concave points       1.300310e-02
worst perimeter            1.238390e-02
worst texture              6.501548e-03
worst radius               5.263158e-03
worst area                 2.786378e-03
mean texture               2.063983e-03
mean area                  1.754386e-03
worst concavity            1.547988e-03
mean concavity             1.341589e-03
mean concave points        1.135191e-03
perimeter error            9.287926e-04
area error                 7.223942e-04
mean radius                5.159959e-04
worst compactness          5.159959e-04
symmetry error             2.063983e-04
concave points error       2.063983e-04
mean smoothness            2.063983e-04
mean compactness           1.031992e-04
concavity error            1.031992e-04
mean fractal dimension     2.221259e-17
texture error              0.000000e+00
mean perimeter             0.000000e+00
worst fractal dimension    0.000000e+00
mean symmetry              0.000000e+00
smoothness error          -1.031992e-04
compactness error         -2.063983e-04
radius error              -3.095975e-04
worst smoothness          -4.127967e-04
worst symmetry            -4.127967e-04
fractal dimension error   -4.127967e-04
dtype: float64

weighted_ensemble_k0_l1
RandomForestClassifierEntr_STACKER_l0    0.800000
LightGBMClassifier_STACKER_l0            0.133333
CatboostClassifier_STACKER_l0            0.066667
dtype: float64
```

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by Innixma** [comment]
@zhanghang1989 Another transient CI error, can we try to solve the HTTP errors?

**Comment by zhanghang1989** [comment]
I am re-triggering the CI and hope it resolves the problem.

File: autogluon/utils/tabular/ml/learner/abstract_learner.py [hunk 1]
```diff
@@ -452,6 +452,23 @@ def leaderboard(self, X=None, y=None, silent=False):
                 print(leaderboard)
         return leaderboard
 
+    # TODO: Should this return a dictionary or pandas Series? Current it returns a Series
```
Comment (jwmueller) [inline autogluon/utils/tabular/ml/learner/abstract_learner.py]: I think sticking with pandas formats for things like leaderboard, feature importance, ensemble model weights is a nice way to go.  But for nested info like from HPO, we should stick with dicts.

File: autogluon/utils/tabular/ml/learner/abstract_learner.py [hunk 2]
```diff
@@ -452,6 +452,23 @@ def leaderboard(self, X=None, y=None, silent=False):
...
+    # TODO: enable_fit_continuation must be set to True to be able to pass X and y as None in this function, otherwise it will error.
+    # Warning: This can take a very, very long time to compute if the data is large and the model is complex.
+    # TODO: Warning: This is not memory safe, large datasets may result in OOM errors during feature importance calculation.
+    # A value of 0.01 means that the objective metric error would be expected to increase by 0.01 if the feature were removed.
```
Comment (jwmueller) [inline autogluon/utils/tabular/ml/learner/abstract_learner.py]: I don't think this is how these values should be explained in the eventual documentation, because they're often a poor substitute for remove-and-retrain values.  Rather, I think they're more appropriately described as how much the model depends on the feature (A value of 0.01 indicates the metric error is expected to increase by 0.01 when the information regarding this feature's value is hidden from the model).

File: autogluon/utils/tabular/ml/models/abstract/abstract_model.py [hunk 3]
```diff
@@ -228,19 +233,45 @@ def compute_feature_importance(self, X, y, features_to_use=None):
...
 
         if not features_to_use:
             features = X.columns.values
         else:
             features = features_to_use
-        feature_count = len(features)
 
-        model_score_base = self.score(X=X, y=y)
+        feature_importance_quick_dict = self.get_model_feature_importance()
+        # TODO: Also consider banning features with close to 0 importance
```
Comment (jwmueller) [inline autogluon/utils/tabular/ml/models/abstract/abstract_model.py]: I don't think we need to ban features with ~0 importance, as long as: A) we are sorting them based on importance in the presented output, B) we are using an output format (like pandas) that will automatically truncate lengthy outputs when printing them.

File: autogluon/utils/tabular/ml/models/abstract/abstract_model.py [hunk 4]
```diff
@@ -228,19 +233,45 @@ def compute_feature_importance(self, X, y, features_to_use=None):
...
+        feature_importances = feature_importances.sort_values(ascending=False)
 
+        return feature_importances
+
+    # TODO: Consider repeating with different random seeds and averaging to increase confidence
+    # TODO: Optimize this
+    # TODO: Check memory usage to avoid OOM
+    # Compute feature importance via permutation importance
+    # Note: Expensive to compute
+    #  Time to compute is O(predict_time*num_features)
```
Comment (jwmueller) [inline autogluon/utils/tabular/ml/models/abstract/abstract_model.py]: Could also speed this up if need be by subsampling the data used to compute importances.  For instance, when `time_limits` are imposed and it is detected that this operation will take significant chunk of the remaining time otherwise.

File: autogluon/utils/tabular/ml/models/catboost/catboost_model.py [hunk 5]
```diff
@@ -201,3 +201,10 @@ def fit(self, X_train, Y_train, X_test=None, Y_test=None, time_limit=None, **kwa
                 self.model.shrink(ntree_start=0, ntree_end=best_iteration+1)
 
         self.params_trained['iterations'] = self.model.tree_count_
+
+    def get_model_feature_importance(self):
```
Comment (jwmueller) [inline autogluon/utils/tabular/ml/models/catboost/catboost_model.py]: Just want to confirm these importances will be on similar scale to those computed by the other models using `AbstractModel. compute_permutation_importance()`?

File: autogluon/utils/tabular/ml/models/lgb/lgb_model.py [hunk 6]
```diff
@@ -277,3 +277,9 @@ def _set_default_searchspace(self):
         for key in self.nondefault_params: # delete all user-specified hyperparams from the default search space
             _ = def_search_space.pop(key, None)
         self.params.update(def_search_space)
+
+    def get_model_feature_importance(self):
+        feature_names = self.model.feature_name()
+        importances = self.model.feature_importance()
```
Comment (jwmueller) [inline autogluon/utils/tabular/ml/models/lgb/lgb_model.py]: Will these be on similar scale as those computed by the other models using `AbstractModel. compute_permutation_importance()`?

File: autogluon/utils/tabular/ml/models/rf/rf_model.py [hunk 7]
```diff
@@ -124,3 +124,13 @@ def hyperparameter_tune(self, X_train, X_test, Y_train, Y_test, scheduler_option
...
         hpo_models = {self.name: self.path}
         return hpo_models, hpo_model_performances, hpo_results
+
+    def get_model_feature_importance(self):
+        if self.features is None:
+            # TODO: Consider making this raise an exception
+            logger.warning('Warning: get_model_feature_importance called when self.features is None!')
+            return dict()
+        feature_names = self.features
+        importances = self.model.feature_importances_
```
Comment (jwmueller) [inline autogluon/utils/tabular/ml/models/rf/rf_model.py]: Will these be on similar scale as those computed by the other models using AbstractModel. compute_permutation_importance()?

File: autogluon/utils/tabular/ml/trainer/abstract_trainer.py [hunk 8]
```diff
@@ -826,6 +829,38 @@ def _get_dummy_stacker(self, level, use_orig_features=True):
         )
         return dummy_stacker
 
+    # TODO: Add option for raw=True, which uses black-box method to get importances for original features instead of whatever features the model uses.
+    #  This is non-trivial to implement for multi-layer stacking ensembles on the OOF data.
+    def get_feature_importance(self, model, X=None, y=None):
```
Comment (jwmueller) [inline autogluon/utils/tabular/ml/trainer/abstract_trainer.py]: For presenting importances to user, trainer should eventually have a separate feature_importance() method that automatically uses the best model, and even if it's a stacker, only considers the original features for importance scoring.

File: autogluon/utils/tabular/ml/models/catboost/catboost_model.py (hunk 5 above)
Comment (Innixma) [reply autogluon/utils/tabular/ml/models/catboost/catboost_model.py]: They are not, this is currently used to get features which are never used by the models (such as never used in a split). It is impossible to scale them to be identical to permutation importance unfortunately, since it varies by model.

File: autogluon/utils/tabular/ml/models/lgb/lgb_model.py (hunk 6 above)
Comment (Innixma) [reply autogluon/utils/tabular/ml/models/lgb/lgb_model.py]: ditto

File: autogluon/utils/tabular/ml/models/rf/rf_model.py (hunk 7 above)
Comment (Innixma) [reply autogluon/utils/tabular/ml/models/rf/rf_model.py]: ditto

**Comment by Innixma** [comment]
@zhanghang1989 CI failed in same place again, there is some permissions error with a file during the doc build.

**Comment by zhanghang1989** [comment]
Is the failure relevant to the code change? I can have a careful look later today.

**Comment by Innixma** [comment]
The error is not related to the code changes, it is coming from image classification.

**Comment by szha** [comment]
Job PR-351-3 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-351/3/index.html

File: autogluon/utils/tabular/ml/trainer/abstract_trainer.py (hunk 8 above)
Comment (Innixma) [reply autogluon/utils/tabular/ml/trainer/abstract_trainer.py]: Agreed

File: autogluon/utils/tabular/ml/models/abstract/abstract_model.py (hunk 4 above)
Comment (Innixma) [reply autogluon/utils/tabular/ml/models/abstract/abstract_model.py]: Yep, I already cap it at 10k rows, although thats fairly arbitrary.

File: autogluon/utils/tabular/ml/learner/abstract_learner.py (hunk 2 above)
Comment (Innixma) [reply autogluon/utils/tabular/ml/learner/abstract_learner.py]: Agreed, will clean the descriptions when surfaced to users.

File: autogluon/utils/tabular/ml/learner/abstract_learner.py (hunk 1 above)
Comment (Innixma) [reply autogluon/utils/tabular/ml/learner/abstract_learner.py]: Sounds good
